### PR TITLE
set `content-security-policy`, `x-content-type-options`, and `x-frame-options` headers for console assets

### PR DESCRIPTION
**Pull request by iliana**
This is the Nexus side to https://github.com/oxidecomputer/console/pull/2142, which needs to land before this lands. See also https://github.com/oxidecomputer/product-assurance/issues/50.

[`content-security-policy`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) tells a web browser what type of content, and from which origins, may be loaded on a page. The primary use is to help guard against cross-site scripting attacks and [other kinds of novel attacks on web applications](https://css-tricks.com/css-keylogger/).

[`x-content-type-options: nosniff`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options) tells the web browser to disallow content sniffing that can cause a browser to decide that responses with non-executable content types (e.g. `image/png`) can in fact be used as executable content types (e.g. `text/javascript`). This needs to be set for all console assets.

[`x-frame-options: DENY`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options) disallows embedding the console within another page, which helps to prevent click-jacking attacks. (This is obsoleted by the `frame-ancestors 'none'` CSP directive, but no harm in adding it.)

`content-security-policy` only needs to be set for the console index page, but there's no harm in setting it for the console assets as well.

As part of this change I did some refactoring:
- The common code between the `asset` function and the `serve_console_index` function are now in a single common function. This allows us to ship a gzip-compressed console index in the future.
- Assets are now streamed instead of read completely into memory.
- I removed the dependency on `mime_guess`; we only have a small list of file extensions we're willing to serve, so it doesn't make sense to compile a huge list of content types we'll never use into Nexus.

There may be other headers from https://owasp.org/www-project-secure-headers/ (see the Best Practices tab) that we want but these are probably the most urgent.

### CODE REVIEW

File: nexus/src/external_api/console_api.rs [hunk 1]
```diff
@@ -718,113 +731,135 @@ fn with_gz_ext(path: &Utf8Path) -> Utf8PathBuf {
...
+    rqctx: RequestContext<Arc<ServerContext>>,
+    path_params: Path<RestPathParam>,
+) -> Result<Response<Body>, HttpError> {
+    // asset URLs contain hashes, so cache for 1 year
+    const CACHE_CONTROL: HeaderValue =
+        HeaderValue::from_static("max-age=31536000, immutable");
 
-    Ok(resp.body(file_contents.into())?)
+    let mut path = Utf8PathBuf::from("assets");
+    path.extend(path_params.into_inner().path);
```
[the remark this answers is not in the document]
Comment (iliana) [reply nexus/src/external_api/console_api.rs:846]: I think I had assumed `find_file` does that, but rereading it I don't think it does (it guards against following symlinks). I will add checks to that function and write a test.

I'm debating whether it makes more sense to bail on any path segment that contains `..`, or if we should `std::path::Path::canonicalize` every path as it comes in and bail if it isn't inside `static_dir`.

[the remark this answers is not in the document]
Comment (david-crespo) [reply nexus/src/external_api/console_api.rs:846]: I also thought it did. I don't a huge problem with taking the easy route and telling any `..` to go away.

[the remark this answers is not in the document]
Comment (iliana) [reply nexus/src/external_api/console_api.rs:846]: Good news, Dropshot already does this. https://github.com/oxidecomputer/dropshot/blob/28c63c7f9d3b81729c0b0343640fb815877cdfab/dropshot/src/router.rs#L636-L687

I'm going to add a test on the Omicron side to ensure that remains the case.

File: nexus/src/external_api/console_api.rs [hunk 2]
```diff
@@ -718,115 +731,147 @@ fn with_gz_ext(path: &Utf8Path) -> Utf8PathBuf {
...
-    let content_type = path.file_name().map_or("text/plain", |f| {
-        mime_guess::from_path(f).first_raw().unwrap_or("text/plain")
-    });
-
-    let mut resp = Response::builder()
-        .status(StatusCode::OK)
-        .header(http::header::CONTENT_TYPE, content_type)
-        .header(http::header::CACHE_CONTROL, "max-age=31536000, immutable"); // 1 year
+    let stream = FramedRead::new(file, BytesCodec::new());
+    let body = Body::wrap_stream(stream);
```
Comment (david-crespo) [inline nexus/src/external_api/console_api.rs:828]: Does this mean we don't have to read the entire file into memory at once?

**Review by david-crespo — APPROVED** [review]
All good

File: nexus/src/external_api/console_api.rs (hunk 2 above)
Comment (iliana) [reply nexus/src/external_api/console_api.rs:828]: Yep. That may have been leading to some request latency with the larger bundles, although not sure if it actually was in practice.

**Comment by iliana** [comment]
#5065 pulls in the CSP changes on the console side, so moving out of draft.

File: nexus/src/external_api/console_api.rs (hunk 2 above)
Comment (david-crespo) [reply nexus/src/external_api/console_api.rs:828]: Seems like probably not (this is colo, dogfood looks a bit different due to VPN I assume), but might as well do it the right way.

image

**Pull request merged**